modules/collect/src/main/java/com/opengamma/collect/id/LinkResolver.java

```diff
@@ -20,16 +24,78 @@
   /**
    * Resolves the supplied link, returning the realized target of the link.
    * <p>
-   * Resolution will be triggered by a call to {@link Link#resolve(LinkResolver)}.
    * The implementation of this interface may perform any thread-safe action to obtain
    * the link target. Typically this will involve accessing an underlying data store.
    * If the link cannot be resolved then a {@code LinkResolutionException} will be thrown.
+   * <p>
+   * The type is expressed as a standard {@link Class} object.
+   *
+   * @param <T>  the type of the target of the link
+   * @param identifier  the identifier to be resolved
+   * @param targetType  the target type of the link
+   * @return the resolved target of the link
+   * @throws LinkResolutionException if the link cannot be resolved
+   */
+  public default <T extends IdentifiableBean> T resolve(StandardId identifier, Class<T> targetType) {
+    return resolve(identifier, TypeToken.of(targetType));
+  }
+
+  /**
+   * Resolves the supplied link, returning the realized target of the link.
+   * <p>
+   * The implementation of this interface may perform any thread-safe action to obtain
+   * the link target. Typically this will involve accessing an underlying data store.
+   * If the link cannot be resolved then a {@code LinkResolutionException} will be thrown.
+   * <p>
+   * The type is expressed as a {@link TypeToken}, which allows types like
+   * {@code Trade<Swap>} to be expressed:
+   * <p>
+   * <pre>{@code
+   *  new TypeToken<Trade<Swap>>() {};
+   * }</pre>
    *
    * @param <T>  the type of the target of the link
-   * @param link  the link to be resolved
+   * @param identifier  the identifier to be resolved
+   * @param targetType  the target type of the link
    * @return the resolved target of the link
    * @throws LinkResolutionException if the link cannot be resolved
    */
-  public abstract <T extends IdentifiableBean> T resolve(ResolvableLink<T> link);
+  public abstract <T extends IdentifiableBean> T resolve(StandardId identifier, TypeToken<T> targetType);
+
+  //-------------------------------------------------------------------------
+  /**
+   * Resolves all the links within one property of a bean.
+   * <p>
+   * This takes the specified bean and replaces the target object.
+   * The target must be a property of the bean and the update function must be able to replace the target.
+   * The update must return a new bean, leaving the original unaltered.
+   * <p>
+   * If the target is not resolvable, or resolving the target is already resolved,
```

Comment:
or resolving the target is already resolved -> or the target is already resolved